src/query_plan/fetch_dependency_graph_processor.rs

```diff
@@ -413,3 +419,30 @@ fn flat_wrap_nodes(
         NodeKind::Sequence => PlanNode::Sequence(Arc::new(SequenceNode { nodes })),
     })
 }
+
+impl RebasedFragments {
+    pub(crate) fn new(query_fragments: NamedFragments) -> Self {
+        Self {
+            query_fragments,
+            by_subgraph: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn for_subgraph(
+        &mut self,
+        subgraph_name: NodeStr,
+        subgraph_schema: &ValidFederationSchema,
+    ) -> Option<&NamedFragments> {
+        self.by_subgraph
+            .entry(subgraph_name)
+            .or_insert_with(|| Self::rebase_on(&self.query_fragments, subgraph_schema))
+            .as_ref()
+    }
+
+    fn rebase_on(
+        query_fragments: &NamedFragments,
+        subgraph_schema: &ValidFederationSchema,
+    ) -> Option<NamedFragments> {
+        todo!() // TODO: port JS `NamedFragments.rebaseOn` from `operations.ts`
```

Comment:
this will be implemented in https://github.com/apollographql/federation-next/pull/239